src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,111 @@
+            var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
+
+            await CreateAnalyzer(syntaxFacts, semanticModel)
+                .ComputeRefactoringsAsync(context).ConfigureAwait(false);
+        }
+
+        protected abstract IAnalyzer CreateAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel);
+
+        protected abstract class Analyzer<TSource, TDestination> : IAnalyzer
+            where TSource : SyntaxNode
+            where TDestination : SyntaxNode
+        {
+            protected readonly ISyntaxFactsService _syntaxFacts;
+            protected readonly SemanticModel _semanticModel;
+
+            public Analyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+            {
+                _syntaxFacts = syntaxFacts;
+                _semanticModel = semanticModel;
+            }
+
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var sourceNode = FindNodeToRefactor(root, context);
+                if (sourceNode == null)
+                {
+                    return;
+                }
+
+                if (sourceNode.ContainsDiagnostics)
+                {
+                    return;
+                }
+
+                var destinationNode = Convert(sourceNode);
+                if (destinationNode == null)
+                {
+                    return;
+                }
+
+                if (!Validate(sourceNode, destinationNode, context.CancellationToken))
+                {
+                    return;
+                }
+
+                context.RegisterRefactoring(new MyCodeAction(Title, c =>
+                    UpdateDocumentAsync(root, document, sourceNode, destinationNode)));
+            }
+
+            protected abstract TSource FindNodeToRefactor(SyntaxNode root, CodeRefactoringContext context);
+
+            protected abstract TDestination Convert(TSource source);
+
+            protected abstract bool Validate(TSource source, TDestination destination, CancellationToken cancellationToken);
+
+            protected abstract string Title { get;  }
+
+            // TODO probably protected virtual
+            private Task<Document> UpdateDocumentAsync(
+                SyntaxNode root,
+                Document document,
+                TSource sourceNode,
+                TDestination destinationNode)
+            {
+                root = root.ReplaceNode(root.FindNode(sourceNode.Span), destinationNode);
+                return Task.FromResult(document.WithSyntaxRoot(root));
+            }
+        }
+
+        protected interface IAnalyzer
+        {
+            Task ComputeRefactoringsAsync(CodeRefactoringContext context);
```

Comment:
i don't understand why we need this type at all.  We're already in an abstract type, so why not just use an abstract method on it?

Reply:
Now I have to restore the IAnalyzer class to share semanticModel and cancellationToken within the traverse algorithm.

---
In reply to: [167387922](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387922) [](ancestors = 167387922)